src/flatten.jl

```diff
@@ -0,0 +1,76 @@
+macro register(dist)
+    return quote
+        DistributionsAD.eval(getexpr($(esc(dist))))
+        DistributionsAD.toflatten(::$(esc(dist))) = true
+    end
+end
+function getexpr(Tdist)
+    x = gensym()
+    fnames = fieldnames(Tdist)
+    flattened_args = Expr(:tuple, [:(dist.$f) for f in fnames]...)
+    func = Expr(:->, 
+                Expr(:tuple, fnames..., x), 
+                Expr(:block,
+                    Expr(:call, :logpdf,
+                        Expr(:call, :($Tdist), fnames...),
+                        x,
+                    )
+                )
+            )
+    return :(flatten(dist::$Tdist) = ($func, $flattened_args))
+end
+const flattened_dists = [   Bernoulli,
+                            BetaBinomial,
+                            Binomial,
+                            Geometric,
+                            NegativeBinomial,
+                            Poisson,
+                            Skellam,
+                            PoissonBinomial,
+                            Arcsine,
+                            Beta,
+                            BetaPrime,
+                            Biweight,
+                            Cauchy,
+                            Chernoff,
+                            Chi,
+                            Chisq,
+                            Cosine,
+                            Epanechnikov,
+                            Erlang,
+                            Exponential,
+                            FDist,
+                            Frechet,
+                            Gamma,
+                            GeneralizedExtremeValue,
+                            GeneralizedPareto,
+                            Gumbel,
+                            InverseGamma,
+                            InverseGaussian,
+                            Kolmogorov,
+                            Laplace,
+                            Levy,
+                            LocationScale,
+                            Logistic,
+                            LogitNormal,
+                            LogNormal,
+                            Normal,
+                            NormalCanon,
+                            NormalInverseGaussian,
+                            Pareto,
+                            PGeneralizedGaussian,
+                            Rayleigh,
+                            SymTriangularDist,
+                            TDist,
+                            TriangularDist,
+                            Triweight,
+                            Categorical,
+                            Truncated,
+                        ]
+for T in flattened_dists
+    @eval toflatten(::$T) = true
+end
+toflatten(::Distribution) = false
+for T in flattened_dists
+    eval(getexpr(T))
+end
```

Comment:
I'm a bit confused about what you want to achieve with the implementation of `flatten` (see also my comment below). It seems also annoying that one would have to implement it for all other possible future types that are not listed here explicitly (i.e., it is not sufficient just to define the trait `toflatten(::T) = true` but one also has to add an explicit definition of `flatten`). Moreover, isn't something like
```julia
flatten(dist, x) = (x, args...) -> logpdf(typeof(dist)(args...), x), getfield.(Ref(dist), fieldnames(typeof(dist)))
```
(the only different would that the position of `args...` and `x` is switched in the function).

Reply:
Yes I don't like this part of the code. See the reason above.

Reply:
I will consider this resolved for lack of a better alternative.